// @lbdudc/sensor-dsl/src/lib/SensorGrammarVisitor.js
// Generated from grammar/SensorGrammar.g4 by ANTLR 4.13.2
// jshint ignore: start
import antlr4 from 'antlr4';

// This class defines a complete generic visitor for a parse tree produced by SensorGrammarParser.

export default class SensorGrammarVisitor extends antlr4.tree.ParseTreeVisitor {

	// Visit a parse tree produced by SensorGrammarParser#parse.
	visitParse(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by SensorGrammarParser#sentence.
	visitSentence(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by SensorGrammarParser#createStatement.
	visitCreateStatement(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by SensorGrammarParser#createProduct.
	visitCreateProduct(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by SensorGrammarParser#createSensorGroup.
	visitCreateSensorGroup(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by SensorGrammarParser#createRange.
	visitCreateRange(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by SensorGrammarParser#rangeProperty.
	visitRangeProperty(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by SensorGrammarParser#rangeNumber.
	visitRangeNumber(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by SensorGrammarParser#createDimension.
	visitCreateDimension(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by SensorGrammarParser#createSpatialDimension.
	visitCreateSpatialDimension(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by SensorGrammarParser#createCategoricalDimension.
	visitCreateCategoricalDimension(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by SensorGrammarParser#createParentDimension.
	visitCreateParentDimension(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by SensorGrammarParser#createDimensionProperties.
	visitCreateDimensionProperties(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by SensorGrammarParser#dimPropertyDefinition.
	visitDimPropertyDefinition(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by SensorGrammarParser#createSensor.
	visitCreateSensor(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by SensorGrammarParser#createSensorProperties.
	visitCreateSensorProperties(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by SensorGrammarParser#sensorPropertyDefinition.
	visitSensorPropertyDefinition(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by SensorGrammarParser#createSensorMeasurementData.
	visitCreateSensorMeasurementData(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by SensorGrammarParser#createMeasurementProperty.
	visitCreateMeasurementProperty(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by SensorGrammarParser#addSpatialDimensionToSensor.
	visitAddSpatialDimensionToSensor(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by SensorGrammarParser#addCategoricalDimensionToSensor.
	visitAddCategoricalDimensionToSensor(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by SensorGrammarParser#addBBXToSensor.
	visitAddBBXToSensor(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by SensorGrammarParser#setDeployment.
	visitSetDeployment(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by SensorGrammarParser#deploymentProperty.
	visitDeploymentProperty(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by SensorGrammarParser#srid.
	visitSrid(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by SensorGrammarParser#identifier.
	visitIdentifier(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by SensorGrammarParser#text.
	visitText(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by SensorGrammarParser#hexColor.
	visitHexColor(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by SensorGrammarParser#floatNumber.
	visitFloatNumber(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by SensorGrammarParser#negativeFloat.
	visitNegativeFloat(ctx) {
	  return this.visitChildren(ctx);
	}


	// Visit a parse tree produced by SensorGrammarParser#coordinate.
	visitCoordinate(ctx) {
	  return this.visitChildren(ctx);
	}



}
