// @lbdudc/sensor-dsl/src/lib/SensorGrammarListener.js
// Generated from grammar/SensorGrammar.g4 by ANTLR 4.13.2
// jshint ignore: start
import antlr4 from 'antlr4';

// This class defines a complete listener for a parse tree produced by SensorGrammarParser.
export default class SensorGrammarListener extends antlr4.tree.ParseTreeListener {

	// Enter a parse tree produced by SensorGrammarParser#parse.
	enterParse(ctx) {
	}

	// Exit a parse tree produced by SensorGrammarParser#parse.
	exitParse(ctx) {
	}


	// Enter a parse tree produced by SensorGrammarParser#sentence.
	enterSentence(ctx) {
	}

	// Exit a parse tree produced by SensorGrammarParser#sentence.
	exitSentence(ctx) {
	}


	// Enter a parse tree produced by SensorGrammarParser#createStatement.
	enterCreateStatement(ctx) {
	}

	// Exit a parse tree produced by SensorGrammarParser#createStatement.
	exitCreateStatement(ctx) {
	}


	// Enter a parse tree produced by SensorGrammarParser#createProduct.
	enterCreateProduct(ctx) {
	}

	// Exit a parse tree produced by SensorGrammarParser#createProduct.
	exitCreateProduct(ctx) {
	}


	// Enter a parse tree produced by SensorGrammarParser#createSensorGroup.
	enterCreateSensorGroup(ctx) {
	}

	// Exit a parse tree produced by SensorGrammarParser#createSensorGroup.
	exitCreateSensorGroup(ctx) {
	}


	// Enter a parse tree produced by SensorGrammarParser#createRange.
	enterCreateRange(ctx) {
	}

	// Exit a parse tree produced by SensorGrammarParser#createRange.
	exitCreateRange(ctx) {
	}


	// Enter a parse tree produced by SensorGrammarParser#rangeProperty.
	enterRangeProperty(ctx) {
	}

	// Exit a parse tree produced by SensorGrammarParser#rangeProperty.
	exitRangeProperty(ctx) {
	}


	// Enter a parse tree produced by SensorGrammarParser#rangeNumber.
	enterRangeNumber(ctx) {
	}

	// Exit a parse tree produced by SensorGrammarParser#rangeNumber.
	exitRangeNumber(ctx) {
	}


	// Enter a parse tree produced by SensorGrammarParser#createDimension.
	enterCreateDimension(ctx) {
	}

	// Exit a parse tree produced by SensorGrammarParser#createDimension.
	exitCreateDimension(ctx) {
	}


	// Enter a parse tree produced by SensorGrammarParser#createSpatialDimension.
	enterCreateSpatialDimension(ctx) {
	}

	// Exit a parse tree produced by SensorGrammarParser#createSpatialDimension.
	exitCreateSpatialDimension(ctx) {
	}


	// Enter a parse tree produced by SensorGrammarParser#createCategoricalDimension.
	enterCreateCategoricalDimension(ctx) {
	}

	// Exit a parse tree produced by SensorGrammarParser#createCategoricalDimension.
	exitCreateCategoricalDimension(ctx) {
	}


	// Enter a parse tree produced by SensorGrammarParser#createParentDimension.
	enterCreateParentDimension(ctx) {
	}

	// Exit a parse tree produced by SensorGrammarParser#createParentDimension.
	exitCreateParentDimension(ctx) {
	}


	// Enter a parse tree produced by SensorGrammarParser#createDimensionProperties.
	enterCreateDimensionProperties(ctx) {
	}

	// Exit a parse tree produced by SensorGrammarParser#createDimensionProperties.
	exitCreateDimensionProperties(ctx) {
	}


	// Enter a parse tree produced by SensorGrammarParser#dimPropertyDefinition.
	enterDimPropertyDefinition(ctx) {
	}

	// Exit a parse tree produced by SensorGrammarParser#dimPropertyDefinition.
	exitDimPropertyDefinition(ctx) {
	}


	// Enter a parse tree produced by SensorGrammarParser#createSensor.
	enterCreateSensor(ctx) {
	}

	// Exit a parse tree produced by SensorGrammarParser#createSensor.
	exitCreateSensor(ctx) {
	}


	// Enter a parse tree produced by SensorGrammarParser#createSensorProperties.
	enterCreateSensorProperties(ctx) {
	}

	// Exit a parse tree produced by SensorGrammarParser#createSensorProperties.
	exitCreateSensorProperties(ctx) {
	}


	// Enter a parse tree produced by SensorGrammarParser#sensorPropertyDefinition.
	enterSensorPropertyDefinition(ctx) {
	}

	// Exit a parse tree produced by SensorGrammarParser#sensorPropertyDefinition.
	exitSensorPropertyDefinition(ctx) {
	}


	// Enter a parse tree produced by SensorGrammarParser#createSensorMeasurementData.
	enterCreateSensorMeasurementData(ctx) {
	}

	// Exit a parse tree produced by SensorGrammarParser#createSensorMeasurementData.
	exitCreateSensorMeasurementData(ctx) {
	}


	// Enter a parse tree produced by SensorGrammarParser#createMeasurementProperty.
	enterCreateMeasurementProperty(ctx) {
	}

	// Exit a parse tree produced by SensorGrammarParser#createMeasurementProperty.
	exitCreateMeasurementProperty(ctx) {
	}


	// Enter a parse tree produced by SensorGrammarParser#addSpatialDimensionToSensor.
	enterAddSpatialDimensionToSensor(ctx) {
	}

	// Exit a parse tree produced by SensorGrammarParser#addSpatialDimensionToSensor.
	exitAddSpatialDimensionToSensor(ctx) {
	}


	// Enter a parse tree produced by SensorGrammarParser#addCategoricalDimensionToSensor.
	enterAddCategoricalDimensionToSensor(ctx) {
	}

	// Exit a parse tree produced by SensorGrammarParser#addCategoricalDimensionToSensor.
	exitAddCategoricalDimensionToSensor(ctx) {
	}


	// Enter a parse tree produced by SensorGrammarParser#addBBXToSensor.
	enterAddBBXToSensor(ctx) {
	}

	// Exit a parse tree produced by SensorGrammarParser#addBBXToSensor.
	exitAddBBXToSensor(ctx) {
	}


	// Enter a parse tree produced by SensorGrammarParser#setDeployment.
	enterSetDeployment(ctx) {
	}

	// Exit a parse tree produced by SensorGrammarParser#setDeployment.
	exitSetDeployment(ctx) {
	}


	// Enter a parse tree produced by SensorGrammarParser#deploymentProperty.
	enterDeploymentProperty(ctx) {
	}

	// Exit a parse tree produced by SensorGrammarParser#deploymentProperty.
	exitDeploymentProperty(ctx) {
	}


	// Enter a parse tree produced by SensorGrammarParser#srid.
	enterSrid(ctx) {
	}

	// Exit a parse tree produced by SensorGrammarParser#srid.
	exitSrid(ctx) {
	}


	// Enter a parse tree produced by SensorGrammarParser#identifier.
	enterIdentifier(ctx) {
	}

	// Exit a parse tree produced by SensorGrammarParser#identifier.
	exitIdentifier(ctx) {
	}


	// Enter a parse tree produced by SensorGrammarParser#text.
	enterText(ctx) {
	}

	// Exit a parse tree produced by SensorGrammarParser#text.
	exitText(ctx) {
	}


	// Enter a parse tree produced by SensorGrammarParser#hexColor.
	enterHexColor(ctx) {
	}

	// Exit a parse tree produced by SensorGrammarParser#hexColor.
	exitHexColor(ctx) {
	}


	// Enter a parse tree produced by SensorGrammarParser#floatNumber.
	enterFloatNumber(ctx) {
	}

	// Exit a parse tree produced by SensorGrammarParser#floatNumber.
	exitFloatNumber(ctx) {
	}


	// Enter a parse tree produced by SensorGrammarParser#negativeFloat.
	enterNegativeFloat(ctx) {
	}

	// Exit a parse tree produced by SensorGrammarParser#negativeFloat.
	exitNegativeFloat(ctx) {
	}


	// Enter a parse tree produced by SensorGrammarParser#coordinate.
	enterCoordinate(ctx) {
	}

	// Exit a parse tree produced by SensorGrammarParser#coordinate.
	exitCoordinate(ctx) {
	}



}
